src/System/Hworker.hs

```diff
@@ -262,12 +328,57 @@ brokenQueue hw = "hworker-broken-" <> hworkerName hw
 failedQueue :: Hworker s t -> ByteString
 failedQueue hw = "hworker-failed-" <> hworkerName hw
 
+batchCounter :: Hworker s t -> BatchId -> ByteString
+batchCounter hw (BatchId batch) =
+  "hworker-batch-" <> hworkerName hw <> ":" <> UUID.toASCIIBytes batch
+
 -- | Adds a job to the queue. Returns whether the operation succeeded.
 queue :: Job s t => Hworker s t -> t -> IO Bool
 queue hw j =
-  do job_id <- UUID.toString <$> UUID.nextRandom
+  do job_id <- UUID.toText <$> UUID.nextRandom
      isRight <$> R.runRedis (hworkerConnection hw)
-                (R.lpush (jobQueue hw) [LB.toStrict $ A.encode (job_id, j)])
+                (R.lpush (jobQueue hw) [LB.toStrict $ A.encode (JobRef job_id Nothing, j)])
+
+-- | Adds a job to the queue, but as part of a particular batch of jobs.
+-- Returns whether the operation succeeded.
+queueBatched :: Job s t => Hworker s t -> t -> BatchId -> IO Bool
+queueBatched hw j batch = do
+  job_id <- UUID.toText <$> UUID.nextRandom
+  R.runRedis (hworkerConnection hw) $ do
+    s <- R.hget (batchCounter hw batch) "status"
+    case s of
+      Left err ->
+        liftIO (hwlog hw err) >> return False
+
+      Right Nothing -> do
+        liftIO $ hwlog hw $ "Batch not found: " <> show batch
+        return False
+
+      Right (Just status) | status == "queueing" -> do
+        result <- R.lpush (jobQueue hw) [LB.toStrict $ A.encode (JobRef job_id (Just batch), j)]
+        void $ R.hincrby (batchCounter hw batch) "queued" 1
```

Comment:
I wonder if this should be in a transaction?

Reply:
If not, counter could be off by one -- is that a big deal? It shouldn't be likely, so if it doesn't cause serious problems, may not matter...